### ISSUE TITLE
Doesn't handle paging/sysparm_offset & sysparm_limit

### ISSUE THREAD

**Issue by lingfish** [rbw0/python-servicenow-rest#12]
ServiceNow (at least on our instance) default to 10,000 results, and the module doesn't handle this.  `requests` naturally does via `.links` methods.

I've overridden a method to hackily handle this, but I'm flagging the issue/calling for opinions before doing a PR.

```
        while 'next' in request.links:
            url = request.links['next']['url']
            print(url)

            request = self._session.get(
                url,
#                params=self._format_query(query, fields)
            )

            result.extend(self._handle_response(request, method))

        return result
```

I think, normally, a generator with yield would be the best fit here?  But I don't have the fu to work out dynamically when something is called as a generator, vs expecting a plain "return".

**Comment by rbw0** [comment]
Hey Lingfish,

Great addition.
Sure, a generator could be an idea I suppose.. but that'd introduce some unwanted complexity imo.

I've implemented this in testing in 3e59dc51b602b7ffac556db72f0727c33502f3d7 based on your code.

```
if 'next' in request.links:  # Paged result
    url = request.links['next']['url']
    paged_request = self._session.get(url)
    response['result'].extend(self._handle_response(paged_request, method))
```

What do you think?

Also, `request.links['next']['url']` already contains the full URL with query string and the whole shebang so there's no need to send `params` again. Sending `params` again actually results in the query string getting duplicated (as SN accepts multiple occurrences of a key/value pair and returns whatever was requested), that is, the more pages the longer the URL would become. 

Thanks :)

**Comment by rbw0** [comment]
Oh ye, I added support for setting "default payload" in 3d50be6c9df18f970474d38ee06a458598edcf4d. It's been committed to the master branch.

Usage as follows:
sn.Client('instance_name', 'user_name', 'password', raise_on_empty=True, default_payload={'sysparm_limit: xxx, 'sysparm_offset': xxx, 'what': 'ever'})

Makes testing this a bit easier I suppose :)

**Comment by lingfish** [comment]
Heya, so a few comments.  Your suggested commit; won't the way you're doing it mean the next paged result fetch will only happen once?  This is why I used a while, and re-used the same `request` object, extending the result, so that we stay in `_handle_response` until we're done.

> Sure, a generator could be an idea I suppose.. but that'd introduce some unwanted complexity imo.

Yeah, I don't doubt that, but I think merely extending and extending a result, potentially fetching millions of rows, is the "wrong way".  A generator, I thought (correct me if I'm wrong), was for exactly this purpose -- so that code that may return a bucketload of data could be "pipelined" to a caller, which then would deal with smaller chunks, and re-init the variable they're using, keeping memory use down on either side?

> Also, request.links['next']['url'] already contains the full URL with query string and the whole shebang so there's no need to send params again.

Yep, agree -- that's why I had it commented out (and forgot to remove from this issue), or am I missing something there?

The default payload commit is an awesome idea :)

**Comment by rbw0** [comment]
> Heya, so a few comments. Your suggested commit; won't the way you're doing it mean the next paged result fetch will only happen once? This is why I used a while, and re-used the same request object, extending the result, so that we stay in _handle_response until we're done.

Recursion takes care of that, a next link is returned from each subsequent request.
I've tested it and it works fine.
```
# ./test.py (sysparm_limit set to 100, 459 results):
fetching next: https://xxx.service-now.com/api/now/table/incident?sysparm_query=state%3D2&sysparm_limit=100&sysparm_offset=100
fetching next: https://xxx.service-now.com/api/now/table/incident?sysparm_query=state%3D2&sysparm_limit=100&sysparm_offset=200
fetching next: https://xxx.service-now.com/api/now/table/incident?sysparm_query=state%3D2&sysparm_limit=100&sysparm_offset=300
fetching next: https://xxx.service-now.com/api/now/table/incident?sysparm_query=state%3D2&sysparm_limit=100&sysparm_offset=400
```

> Yeah, I don't doubt that, but I think merely extending and extending a result, potentially fetching millions of rows, is the "wrong way". A generator, I thought (correct me if I'm wrong), was for exactly this purpose -- so that code that may return a bucketload of data could be "pipelined" to a caller, which then would deal with smaller chunks, and re-init the variable they're using, keeping memory use down on either side?

Yes, I agree.
@jcpunk is requesting a small change in #13 to make the code thread safe , I fear implementing a generator would probably make it more difficult to accomplish this.
Could one implementation eliminate the need of the other here? Or perhaps there's another solution that takes care of both issues?

**Comment by rbw0** [comment]
I've created a test version using generators in: 207d2e2c9c98af23ebaba5efcaea7da6aeaef152

It works in Python 2.7 and 3.4.
The generators are chained in get() so the change should be seamless.

@jcpunk - not sure how this will affect your implementation of threads? Guessing it might break :(

Anyway, please test it out and tell me what you think.

**Comment by jcpunk** [comment]
Works for me for now.  The feature makes sense and I'm probably not going to circle back to the more complex threading stuff for a while.

**Comment by lingfish** [comment]
Hi Robert, 207d2e2 looks good thus far in my limited testing!  And of course, it just slots right into existing `for` loop code.  Very nice.  A super basic test also yields (get it!) a number greater than 10,000 too:

```
In [21]: for i in devices:
   ....:     counter += 1
   ....:

In [22]: counter
Out[22]: 19326
```

Regarding threading... I've actually tackled this a different way (as I needed to fetch multiple results simultaneously for speed), and for a learning experience, used [RQ](http://python-rq.org/).  Though the principle of queueing doesn't really fit my project, the fact that it multiprocesses almost as a side-effect is working great for me, and cut ServiceNow fetch times by more than 60%.

**Comment by rbw0** [comment]
There's a project based on requests called grequests (same creator as requests), which uses gevent for concurrency and supports generators natively:
https://github.com/kennethreitz/grequests

I've built a test version of python-servicenow-rest using this library, and it works pretty good.
However, there are some problems:
1) It doesn't offer support for link headers, even if it did, it wouldn't be possible to accomplish concurrency on these requests due to the nature of how link headers works.
2) Implementing grequests without API changes is possible, however it increases code complexity quite a bit.

I'm considering creating a separate project that uses grequests instead, and at the same time take the opportunity to rebuild some of the fundamental parts of this library.

What do you guys think? Would a separate version of python-servicenow-rest with support for concurrency (and generators) be a good idea?

**Comment by rbw0** [comment]
Here's an example of how it can look on the user end:
https://gist.github.com/rbw0/102dc4191a8d520dfd4303b7d685a632

**Comment by rbw0** [comment]
Hello,

I've created a new library as a replacement for this one called pysnow (as in "python service now"), I'll reference to it in the readme once I've added it to pypi.

Much of the functionality has been kept and some added, i.e. generators (the topic of this discussion).
I decided not to implement non-blocking / async functionality as this doesn't really make sense in the library, I'll leave this up to the user instead.

Anyway, it's available here if you wanna check it out:
https://github.com/rbw0/pysnow

I'm closing this issue. Hope you'll use the new library! :)

**Comment by lingfish** [comment]
Sorry I lagged on my reply.

I agree with you.  I don't think the gevents idea or async support is your issue to solve -- this is why I went with rq.  K.I.S.S. and all that.

Why the new project though?  Can't this one just be renamed, keeping all history etc?

**Comment by rbw0** [comment]
Hey,

Here are some reasons to why.
1) I wasn't happy with the code structure in general. Also, some of the property and method names were misleading / confusing.
2) I wanted a simpler / cleaner user API
3) Thread safety; python-servicenow-rest isn't thread-safe. This can be fixed but that'd mean either a) replace the non-thread-safe code and affect the user API, or b) add code which can be used if you want thread-safety and tell the user about deprecated code usage. This would increase complexity and maintenance - I rather see that all code is thread-safe and least amount of maintenance as possible.
4) I wanted to document the project properly, but I didn't want to document code that I wasn't fully satisfied with.
5) The projected needed a new file structure, a package with just a single module doesn't make sense.

To summarize; I felt that many small things needed to be done, some were ideas of my own and some from you guys (i.e. thread-safety and generators). Instead of cluttering up the code by implementing new, (possibly) redundant code with deprecation warnings or making API-affecting changes, I decided to create a new project with most stuff taken from this project, and leave this one as-is.
